forwardproxy.go

```diff
@@ -395,21 +398,58 @@ func (fp *ForwardProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int,
 			r.Header.Add("Via", strconv.Itoa(r.ProtoMajor)+"."+strconv.Itoa(r.ProtoMinor)+" caddy")
 		}
 
-		if fp.responseTimeout != nil {
-			targetConn.SetDeadline(time.Now().Add(*fp.responseTimeout))
-		}
-
+		var err error
 		var response *http.Response
-		err = r.Write(targetConn)
-		if err != nil {
-			return http.StatusBadGateway, errors.New("failed to write http request: " + err.Error())
+		if fp.upstream == nil {
+			// non-upstream request uses httpTransport to reuse connections
+			if r.Body != nil &&
+				(r.Method == "GET" || r.Method == "HEAD" || r.Method == "OPTIONS" || r.Method == " TRACE") {
+				// make sure request is idempotent and could be retried by saving the Body
+				rBodyBuf, err := ioutil.ReadAll(r.Body)
+				if err != nil {
+					return http.StatusBadRequest, errors.New("failed to read request Body: " + err.Error())
+				}
+				r.GetBody = func() (io.ReadCloser, error) {
+					return ioutil.NopCloser(bytes.NewReader(rBodyBuf)), nil
+				}
+				r.Body, _ = r.GetBody()
+			}
+			response, err = fp.httpTransport.RoundTrip(r)
+		} else {
+			// Upstream requests don't interact well with Transport: connections could always be
+			// reused, but Transport thinks they go to different Hosts, so it spawns tons of
+			// useless connections. Just use dialContext, which will multiplex via single connection
```

Comment:
Will `dialContext` multiplex if the upstream connection is HTTP/1.1?  The comment could be clearer.

Reply:
Fixed